core/matrix/csr.cpp

```diff
@@ -92,8 +110,25 @@ void Csr<ValueType, IndexType>::apply_impl(const LinOp *alpha, const LinOp *b,
                                            const LinOp *beta, LinOp *x) const
 {
     using Dense = Dense<ValueType>;
-    this->get_executor()->run(csr::make_advanced_spmv(
-        as<Dense>(alpha), this, as<Dense>(b), as<Dense>(beta), as<Dense>(x)));
+    using TCsr = Csr<ValueType, IndexType>;
+    if (auto b_csr = dynamic_cast<const TCsr *>(b)) {
+        auto exec = this->get_executor();
+        Array<IndexType> x_rows(exec);
+        Array<IndexType> x_cols(exec);
+        Array<ValueType> x_vals(exec);
+        auto x_csr = as<TCsr>(x);
```

Comment:
If x is not csr here, x_csr is new pointer, right?
new_x ->move_to(x_csr) will not change anything

Reply:
No, if `x` is not CSR, it throws a `NotSupported` exception.
See [the implementation of `as`](https://github.com/ginkgo-project/ginkgo/blob/develop/include/ginkgo/core/base/utils.hpp#L288) for details.

Reply:
I got it. Thanks